luks2/reader.go

```diff
@@ -64,13 +77,63 @@ func (s *storageContainerReadWriterImpl) ensureKeyslotNames() error {
 		keyslots[name] = &keyslotInfoImpl{
 			keyslotType: secboot.KeyslotTypeRecovery,
 			keyslotName: name,
+			keyslotId:   internal_luks2.AnySlot, // use AnySlot to indicate we haven't filled this KeyslotInfo yet.
 		}
 	}
 
 	s.keyslots = keyslots
 	return nil
 }
 
+func (s *storageContainerReadWriterImpl) ensureKeyslotInfo(ctx context.Context, name string) error {
+	if err := s.ensureKeyslotNames(); err != nil {
+		return err
+	}
+
+	info, exists := s.keyslots[name]
+	if !exists {
+		return secboot.ErrKeyslotNotFound
+	}
+
+	if info.keyslotId != internal_luks2.AnySlot {
+		// We already have everything for this keyslot.
+		return nil
+	}
+
+	switch info.keyslotType {
+	case secboot.KeyslotTypeRecovery:
+		// The existing secboot API doesn't expose the LUKS2 keyslot ID for recovery
+		// keys, so we use the luksview package directly, via a bit of an abstraction
+		// so that it can be mocked in unit tests. This will all be cleaned up
+		// eventually once all of this functionality is implemented natively in this
+		// package.
+		//
+		// The new unlocking API will test a recovery key separately against each
+		// individual recovery keyslot, so we know which keyslot is used for unlocking.
+		view, err := newLuksView(ctx, s.container.Path())
+		if err != nil {
+			return fmt.Errorf("cannot obtain new luksview.View: %w", err)
+		}
+		token, _, inUse := view.TokenByName(name)
+		if !inUse {
+			return fmt.Errorf("no metadata for keyslot %q", name)
+		}
+		info.keyslotId = token.Keyslots()[0] // luksview guarantees there is always 1 keyslot here.
+	case secboot.KeyslotTypePlatform:
+		r, err := luks2Ops.NewKeyDataReader(s.container.Path(), name)
+		if err != nil {
+			return fmt.Errorf("cannot obtain reader for %q: %w", name, err)
+		}
+		info.keyslotId = r.KeyslotID()
+		info.keyslotPriority = r.Priority()
+		info.keyslotData = r
```

Comment:
how does this will notice updates if it is cached if we don't have yet a write side?